Exact chainrules derivatives for `beta_inc` and `beta_inc_inv`

lrnv opened the pull request:
## Summary

This PR adds ChainRules coverage for the regularized incomplete beta function and its inverse.

- Adds analytic `frule`/`rrule` definitions for:
  - `beta_inc(a, b, x) -> (p, q)`
  - `beta_inc(a, b, x, y)`, with `x + y = 1`
  - `beta_inc_inv(a, b, p) -> (x, 1 - x)`
  - `beta_inc_inv(a, b, p, q)`, with `p + q = 1`
- Uses symmetric tangent/cotangent splits for the redundant four-argument parametrizations.
- Supports integer inputs through promotion and supports `Float16`/`Float32` by computing the differentiated continued fraction in `Float64` and converting back.
- Handles endpoints and degenerate shape parameters explicitly.
- Adds extensive tests for forward and reverse rules, endpoint behavior, type stability, allocations, promotion, non-convergence, and derivatives along the four-argument constraints.

## Implementation

The parameter derivatives of `beta_inc` are computed by differentiating the continued-fraction method described by Boik & Robinson-Cox (1999). This is **not a translation of the S+ source code**: the implementation is independently adapted from [arzwa/IncBetaDer](https://github.com/arzwa/IncBetaDer), now MIT-licensed, whose author also [explicitly permitted reuse](https://github.com/JuliaMath/SpecialFunctions.jl/pull/506#issuecomment-3369042219).

The continued fraction exits as soon as its values converge. Its upper bound is `maxapp = 10_000`, rather than the historical value of 200, because difficult nearly symmetric cases with large shape parameters require substantially more approximants. If convergence is not reached, the helper emits a rate-limited warning and returns `NaN` derivatives instead of silently returning an inaccurate gradient.

Small recurrence helpers used only once have been integrated into the loop. The remaining shared helper stays `@inline` because it is called four times per iteration and benchmarking showed a material performance benefit.

## Motivation

This allows automatic differentiation through `beta_inc` and `beta_inc_inv`. Missing derivatives affect, among other use cases:

- beta quantiles and beta-distribution marginals;
- Student-t and F distribution workflows built on incomplete beta functions;
- fitting `Distributions.MvTDist`;
- applications in packages such as Copulas.jl;
- the use case discussed in [this Julia Discourse thread](https://discourse.julialang.org/t/gaussian-copula-priors-in-turing-auto-differentiation-error-with-beta-quantile-function/132779/5).

These rules naturally belong in SpecialFunctions.jl's ChainRulesCore extension.

## Validation

The tests cover a broad grid of `(a, b, x)`, including values close to `x = 0` and `x = 1`, integer and reduced-precision inputs, and the constrained four-argument APIs. Sensitive large-parameter regression cases include `a = b = 10^6` and `a = b = 10^8` at `x = 0.5`; these exercise cases that need more than 200 and, for `10^8`, thousands of approximants.

The full SpecialFunctions.jl test suite passes, and the new code is covered by the test suite.

## Reference

> Boik, R. J., & Robinson-Cox, J. F. (1999). Derivatives of the incomplete beta function. *Journal of Statistical Software, 3*(1). https://doi.org/10.18637/jss.v003.i01

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,435 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+# Note on incomplete beta derivatives implementation
+# --------------------------------------------------
+# The rules for the regularized incomplete beta I_x(a,b) and its inverse are
+# implemented using a direct translation of the original S-PLUS/MATLAB code by
+# Boik & Robinson-Cox. See:
+#   Boik, R. J., & Robinson-Cox, J. F. (1998).
+#   Derivatives of the incomplete beta function with respect to its parameters.
+#   Computational Statistics & Data Analysis, 27(1), 85–106.
```
Hum... So we cannot use it ? I dont know very much about licences.

asinghvi17 replied:
They link GPL compatible licenses which includes MIT.   The software itself has no license materials so I think we are fine here?

devmotion replied:
> The software itself has no license mentions at all that I could find, even in the source code

To my understanding, no license means all rights are reserved: https://opensource.stackexchange.com/a/1721

lrnv replied:
Lying would be easy, but I did not implement it from scratch from the paper, I literally translated their code (which was much easier to do of course). However I am not sure the code itself was indeed licenced, the JSS website is very recent compared to the time the paper was published, and I do not now if at the time licencing was imposed on authors as thoroughly as it is today.

asinghvi17 replied:
That seems like a reasonable solution, since there is ambiguity on what exactly the license is (unspecified => all rights reserved, GPL or something else 😅)

lrnv replied:
One is emeritus since 15 years, the other one I cannot find an e-mail he might be dead.. This is getting complicated ^^

arzwa commented:
I agree, you can use the code as you wish, I added an MIT license to the repo.

lrnv commented:
I knew it'll be faster 🤣

arzwa commented:
> @asinghvi17, @devmotion: I have switched implementations from the porting of the original code (which was probably not licensed correctly as we discussed yesterday) to a port of [this repository](https://github.com/arzwa/IncBetaDer), which is an independent implementation from @arzwa, unlicensed yet.

Now that I'm reading mentioned discussion, I would like to add that I definitely implemented this using the approach described in the mentioned paper (Boik & Robinson-Cox 1998). As far as I remember, I did not 'translate' the code but implemented the described numerical method 'independently' from their mathematical description (in fact I'm pretty sure I have never seen S-plus code in my life). However, I don't know what that would mean license-wise and whether it's OK to use an MIT license (I have never had to think about those kind of issues before). Any insights on this?

devmotion reviewed (commented):
All the nested functions, multiple assignments on a single line and generally lack of comments make it quite challenging to follow the code. Maybe the implementation should be cleaned up a bit before it should be considered for inclusion in SpecialFunctions.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    zeroT = zero(T)
+
+    # 1) Boundary cases for x
+    x == oneT && return oneT, zeroT, zeroT, zeroT
```
Done.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    minapp = max(5, minapp)
+
+    # 3) Non-boundary path: precompute ∂I/∂x at original (a,b,x) via stable log form
+    dx = exp((a - oneT) * log(x) + (b - oneT) * log1p(-x) - logbeta(a,b))
```
Already tried on the main _Kfun, did not work very well. This form is the most stable we have IMHO.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    # derivatives
+    T = promote_type(float(typeof(a)), float(typeof(b)), float(typeof(x)))
+    _, dIa_, dIb_, dIx_ = _beta_inc_grad(T(a), T(b), T(x))
+    dIa::T = dIa_; dIb::T = dIb_; dIx::T = dIx_
```
Done.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    p, q = beta_inc(a, b, x, y)
+    T = promote_type(float(typeof(a)), float(typeof(b)), float(typeof(x)), float(typeof(y)))
+    _, dIa_, dIb_, dIx_ = _beta_inc_grad(T(a), T(b), T(x))
+    dIa::T = dIa_; dIb::T = dIb_; dIx::T = dIx_
```
Done and reprecuted below.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    ΔbT::T = Δb isa Real ? T(Δb) : zero(T)
+    ΔxT::T = Δx isa Real ? T(Δx) : zero(T)
+    ΔyT::T = Δy isa Real ? T(Δy) : zero(T)
+    Δp = dIa * ΔaT + dIb * ΔbT + dIx * (ΔxT - ΔyT)
```
Done and reprecuted below.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+
+function _bnfun(p::T, q::T, f::T, n::Int) where {T<:AbstractFloat}
+    # b_n coefficient (n ≥ 1) of the continued fraction. Derived for the same CF.
+    x = 2 * (p * f + 2 * q) * n^2 + 2 * (p * f + 2 * q) * (p - 1) * n + p * q * (p - 2 - p * f)
```
Done.

droodman commented:
Sorry, I did not check the tests.
I feel a bit out of my depth here. Based on pattern-matching intuition, it seems that one line should use min() and the other max().

## Separately:
* I found that decorating all the small helper functions with `@inline` halved run-time whether with Float64 or ForwardDiff.Dual.

* On that note, I got this working with ForwardDiff (which in my project is working best for automatic differentiation) by making a couple of changes. I changed all instances of `AbstractFloat` to `Real` so it would accept `ForwardDiff.Dual`s. And I added:
```
using ForwardDiff, ForwardDiffChainRules                                                  
import SpecialFunctions.beta_inc                                                          
@ForwardDiff_frule beta_inc(a::ForwardDiff.Dual, b::ForwardDiff.Dual, x::ForwardDiff.Dual)
```
I assume the middle line is only needed because my copy of your code is outside of SpecialFunctions. I'm just using beta_inc() (actually `cdf(TDist(...)...)`) so I only bothered creating a ForwardDiff chain rule for that function.

Again, since I'm out of my depth, I leave it to you to judge if there's any practical upshot for you.

lrnv commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    ΔaT::T = Δa isa Real ? T(Δa) : zero(T)
+    ΔbT::T = Δb isa Real ? T(Δb) : zero(T)
+    ΔxT::T = Δx isa Real ? T(Δx) : zero(T)
+    Δp = dIa * ΔaT + dIb * ΔbT + dIx * ΔxT
```
Found out what you expected :)

devmotion commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -300,4 +300,431 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+
+function _Kfun(x::T, p::T, q::T) where {T<:AbstractFloat}
+    # K(x;p,q) = x^p (1-x)^{q-1} / (p * B(p,q)) computed in log-space for stability
+    return exp(p * log(x) + (q - 1) * log1p(-x) - log(p) - logbeta(p, q))
```
Sorry, I actually meant `xlogy` and `xlog1py`. Why does it error?

devmotion commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -301,4 +301,413 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+        return ChainRulesCore.NoTangent(), ā, b̄, x̄
+    end
+    return (p, q), beta_inc_pullback
+end
```
Isn't this just

```suggestion
ChainRulesCore.@scalar_rule(
    beta_inc(a::Number, b::Number, x::Number),
    @setup((_, dIa, dIb, dIx) = _beta_inc_grad(map(float, promote(a, b, x))...)),
    (dIa, dIb, dIx),
    (-dIa, -dIb, -dIx),
)
```

I doubt the slightly reduced calculations are worth the additional code complexity, in particular given how easy it is to introduce inefficiencies or bugs or miss special cases (such as special tangent types) in custom implementations compared to how battle tested `@scalar_rule` is.

lrnv replied:
Thanks a lot for that, greatly simplifies the end of the code indeed...

lrnv commented:
Hey @devmotion would you consider giving this another look ?

lrnv commented:
Rebased on master; faillure on pre seems unrelated.

devmotion commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -301,4 +301,351 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+# the auxiliary variable f, the continued-fraction coefficients a_n, b_n, and
+# their partial derivatives w.r.t. p (≡ a) and q (≡ b). See Boik & Robinson-Cox (1999).
+
+@inline function _Kfun(logx::T, log1mx::T, p::T, q::T, logbetapq::T) where {T}
```
Is the `@inline` actually needed? Generally I prefer leaving such decisions to the compiler.

lrnv replied:
Not needed, removed.

droodman replied:
I understand the concern about @inline, but in my experience it has a substantial performance impact here. In my own application, it is ~2X, on a Mac M4 Max or an Intel i9-13900H.

This test doesn't quite give 2X, but it's still a significant difference:

```
using BenchmarkTools
ext = Base.get_extension(SpecialFunctions, :SpecialFunctionsChainRulesCoreExt)
@btime begin
    ext._beta_inc_grad(2.0, 1.0, 0.25)    
    ext._beta_inc_grad(1.0, 2.0, 0.75)    
    ext._beta_inc_grad(1e-6, 2.0, 0.01)   
    ext._beta_inc_grad(2.0, 1e-6, 0.99)  
    ext._beta_inc_grad(1000.0, 1000.0, 0.5)
end
```
With @inline's, on the Intel machine I get
```
3.812 μs (5 allocations: 160 bytes)
```
Without, I get
```
5.467 μs (5 allocations: 160 bytes)
```

If you want I can pinpoint which @inline's matter. If you want to get rid of the @inline's, my inclination would be to literally inline the functions. As an editor of this code, feeling like a guest, I treaded lightly where it didn't sacrifice performance.

lrnv replied:
At the beginning of this PR, these small functions were literally inlined, we extra Ted them for clarity. But clarity cannot weaken the performance imho, so let's put them back in. 

Thanks for caring and looking out !

devmotion commented on ext/SpecialFunctionsChainRulesCoreExt.jl:
```diff
@@ -301,4 +301,351 @@ function ChainRulesCore.rrule(::typeof(besselyx), ν::Number, x::Number)
+    # 0) Previously keyword arguments: 
+    maxapp=200
+    minapp=3
+    ϵ=eps(T)*T(1e4)
```
Where are these values coming from? What are they used for? Can this be documented?

It also seems we silently accept whatever result we obtain after `maxapp` iterations. IMO ideally we would error or at least warn if calculations did not converge/desired accuracy is not me instead of silently returning some value. Or alternatively not hardcoding a maximum number of iterations but increase dynamically (although I think it's safer to set a - large - maximum number to avoid excessive computation times).

lrnv replied:
1) minapp/maxapp represent the maximum order of approximants used. 

2) These three default values (epislon included as 1000*eps) are coming directly from the original paper. However, the paper itself does not motivate more clearly these default choices. 

3) minapp = 3 just prevents accepting a lucky early equality before the continued-fraction recurrence has settled. Since the convergence test checks I, ∂I/∂p, and ∂I/∂q, three approximants is a modest minimum, not a performance concern

4) maxapp = 200 is already very high and should practically never be reached for Float64. The tail swap already enforces the more favorable side of the CDF, and using 1000*eps(T) as an espilon should make tens of approximants already largely overkill. 

The cases where maxapp can plausibly be reached are the hard beta-CDF regimes:
- Very large a,b, especially near the mode/mean, e.g. a ≈ b ≫ 1 and x ≈ a/(a+b). This makes the continued fraction converge slowly while the parameter derivatives are also demanding.
- Very small shape parameters with x close to the difficult boundary after symmetry, especially when derivatives with respect to the tiny parameter are important.
- Highly imbalanced parameters with extreme x, depending on which side the swap sends the computation to.

I added specific test for these hard cases and an explicit warning for when the maxapp is reached, which does not happen in the testing, so it seems that indeed maxapp is already very high.

lrnv replied:
The issue is that this extra potential warning test is taking a lot of time ? Tell me if you want me to take it out, i can basically revert the last commit.

devmotion commented on test/chainrules.jl:
```diff
@@ -176,4 +176,152 @@
+            # Direct derivative checks without FD: ∂I/∂x equals beta pdf
+            pdf = x^(a - 1) * (1 - x)^(b - 1) / beta(a, b)
+            _, Δx = frule((NoTangent(), 0.0, 0.0, 1.0), beta_inc, a, b, x)
+            @test isapprox(Δx[1], pdf; rtol=1e-12, atol=1e-12)
```
Can we avoid specifying `atol`? It's generally hard to reason about appropriate absolute tolerances without knowing the exact order of magnitude of numbers dealing with.

devmotion replied:
```suggestion
            @test Δx[1] ≈ pdf rtol=1e-12
```

devmotion commented on test/chainrules.jl:
```diff
@@ -176,4 +176,152 @@
+            # Symmetry check: ∂I/∂a(a,b,x) = -∂I/∂b(b,a,1-x)
+            _, Δa = frule((NoTangent(), 1.0, 0.0, 0.0), beta_inc, a, b, x)
+            _, Δb_sw = frule((NoTangent(), 0.0, 1.0, 0.0), beta_inc, b, a, 1 - x)
+            @test isapprox(Δa[1], -Δb_sw[1]; rtol=1e-10, atol=1e-12)
```
```suggestion
            @test Δa[1] ≈ -Δb_sw[1] rtol=1e-12
```

devmotion commented on test/chainrules.jl:
```diff
@@ -176,4 +176,152 @@
+            x_inv, _ = beta_inc_inv(a, b, p)
+            # Check primal composition
+            p_roundtrip = first(beta_inc(a, b, x_inv))
+            @test isapprox(p_roundtrip, p; rtol=1e-12, atol=1e-12)
```
```suggestion
            @test p_roundtrip ≈ p rtol=1e-12
```

devmotion commented on test/chainrules.jl:
```diff
@@ -176,4 +176,152 @@
+            # Forward through g then f: dp
+            _, Δx_inv_dp = frule((NoTangent(), 0.0, 0.0, 1.0), beta_inc_inv, a, b, p)
+            _, Δp_from_dp = frule((NoTangent(), 0.0, 0.0, Δx_inv_dp[1]), beta_inc, a, b, x_inv)
+            @test isapprox(Δp_from_dp[1], 1.0; rtol=1e-9, atol=1e-12)
```
```suggestion
            @test Δp_from_dp[1] ≈ 1.0 rtol=1e-12
```

devmotion replied:
Etc below...

devmotion commented on test/chainrules.jl:
```diff
@@ -176,4 +176,152 @@
+            # Forward da
+            _, Δx_inv_da = frule((NoTangent(), 1.0, 0.0, 0.0), beta_inc_inv, a, b, p)
+            _, Δp_from_da = frule((NoTangent(), 1.0, 0.0, Δx_inv_da[1]), beta_inc, a, b, x_inv)
+            @test isapprox(Δp_from_da[1], 0.0; rtol=1e-9, atol=1e-12)
```
```suggestion
            @test iszero(Δp_from_da[1])
```

lrnv commented:
@devmotion Thanks for looking around ! I had troubles with `iszero(-1e-19)` was not working as good as `isapprox(-1e-19, 0.0, atol=1e-16)`, so i kept the last one in two particular places. Otherwise i got rid of all `atol`s as requested. 

Furthermore, as @droodman said, the numerous `inline`  of small functions are here really needed for performance. We could get rid of it by literally inlining them, but that'll be going backward w.r.t. previous discussions. 

Still failling on pre for unrelated reasons. You may take another look.

devmotion reviewed (commented):
Given the size of the diff I consulted Claude to make this review more thorough than I'd have managed by hand.

`_beta_inc_grad` agrees with central differences to 1e-10..1e-13 over a wide range of `(a, b, x)`, including `(500, 3, 0.99)`, `(3, 500, 0.001)` and `(1e-3, 2, 0.01)`, at ~295ns vs ~230ns for the primal. So the continued fraction itself is not the problem — everything below is in the rules and the edge cases.

### The 4-argument rule is off by a factor of 2

`_beta_inc` throws unless `x + y ≈ 1` (`src/beta_inc.jl:766`), so `y` is not a free argument and there is no `∂/∂y` at fixed `x` — perturbing one of them alone just leaves the domain. The only observable derivative is the one along the constraint. For `f(z) = beta_inc(a, b, g(z), 1 - g(z))[1]` the chain rule gives `g'(z) * (∂x - ∂y)`, so whatever the rule does it has to satisfy

```
∂x - ∂y == ∂I/∂x
```

The PR returns `(dIx, -dIx)`, which gives `2 * dIx`. I checked: with `g = logistic(z)` and `g = logistic(-z)`, `a ∈ (0.4, 1.2, 45.0)`, `b ∈ (2.3, 5.0, 16.0)`, `z ∈ (-1.5, 0.0, 0.8)`, both `frule` and `rrule` come out at exactly 2.000x the finite difference in `z` — all 18 combinations.

Since `x` and `y` are a redundant parametrisation I'd split symmetrically, the same way we handle `Symmetric`/`Hermitian` cotangents:

```suggestion
    (dIa, dIb, dIx / 2, -dIx / 2),
    (-dIa, -dIb, -dIx / 2, dIx / 2),
```

With that all 18 cases give 1.000x. `(dIx, ZeroTangent())` satisfies the identity as well but makes one argument a gradient sink, which seems worse in case most of the dependency happens to flow through `y`. Please don't use `NoTangent()` here — `y` is a perfectly differentiable `Number` and it would end up in a slot where `ProjectTo` expects a `Real`.

Worth getting right since `beta_inc(a, b, x, y)` is the form we advertise in `docs/src/functions_overview.md:27`, so it's what people will copy.

The current tests encode the factor 2 (`test/chainrules.jl:316-318` and `340`), so they need to change too. I'd drop the hand-written comparison of the individual partials and test the composition instead — that *is* testable with finite differences, unlike the raw 4-arg form, since `z -> (g(z), 1 - g(z))` never leaves the domain:

```julia
@testset "4-arg beta_inc: derivative along the constraint y = 1 - x" begin
    # ∂/∂x and ∂/∂y are not separately testable since beta_inc requires x + y == 1.
    # What is observable is the derivative along the constraint, ie. for x = g(z),
    # y = 1 - g(z) both modes have to reproduce
    #     d/dz I_{g(z)}(a, b) = pdf(g(z)) * g'(z),
    # ie. the rule has to satisfy ∂x - ∂y == ∂I/∂x, however it splits the two.
    logistic(z) = inv(1 + exp(-z))
    # opposite signs of g' and x mirrored around 1/2, so that sign errors show up and
    # the internal tail swap is exercised in both directions
    gs = (
        (name = "logistic(z)",  g = z ->  logistic(z), g′ = z ->  logistic(z)  * (1 - logistic(z))),
        (name = "logistic(-z)", g = z -> logistic(-z), g′ = z -> -logistic(-z) * (1 - logistic(-z))),
    )
    for gg in gs, a in (0.4, 1.2, 45.0), b in (2.3, 5.0, 16.0), z in (-1.5, 0.0, 0.8)
        x = gg.g(z)
        d = gg.g′(z)
        expected = x^(a - 1) * (1 - x)^(b - 1) / beta(a, b) * d
        @testset "$(gg.name) a=$a b=$b z=$z" begin
            @test all(isapprox.(beta_inc(a, b, x, 1 - x), beta_inc(a, b, x); rtol=1e-12))

            _, Δ = frule((NoTangent(), 0.0, 0.0, d, -d), beta_inc, a, b, x, 1 - x)
            @test isapprox(Δ[1], expected; rtol=1e-12)
            @test isapprox(Δ[2], -expected; rtol=1e-12)

            _, pb = rrule(beta_inc, a, b, x, 1 - x)
            _, ā, b̄, x̄, ȳ = pb((1.0, 0.0))
            @test isapprox(x̄ * d + ȳ * -d, expected; rtol=1e-12)

            _, pb3 = rrule(beta_inc, a, b, x)
            _, ā3, b̄3, _ = pb3((1.0, 0.0))
            @test isapprox(ā, ā3; rtol=1e-12)
            @test isapprox(b̄, b̄3; rtol=1e-12)
        end
    end
end
```

Also `@test isapprox(ȳ4, -x̄3; rtol=1e-1)` on line 337 — that relation is exact by construction and everything around it uses `1e-11`, so I assume `1e-1` is left over from debugging.

### `_beta_inc_grad` returns four values for `x == 1`

```julia
isone(x) && return oneT, zeroT, zeroT, zeroT
```

Every other branch returns three, and all three call sites destructure three. Julia silently drops the extra element, so `dIa` picks up the primal:

```julia
julia> ext._beta_inc_grad(3.0, 1.0, 1.0)
(1.0, 0.0, 0.0, 0.0)

julia> frule((NoTangent(), 1.0, 0.0, 0.0), beta_inc, 1.2, 2.3, 1.0)[2]
Tangent{Tuple{Float64, Float64}}(1.0, -1.0)   # should be (0.0, 0.0)
```

The `x == 0` branch has the same arity problem, it just happens to be all zeros. Though `∂I/∂x` at `x == 0` is `b` for `a == 1` and `Inf` for `a < 1`, not `0`, so that branch isn't right either.

Independently of the wrong value this costs an allocation on *every* call, because the two arities widen the return type:

```julia
julia> Base.return_types(ext._beta_inc_grad, Tuple{Float64,Float64,Float64})
1-element Vector{Any}:
 Union{Tuple{Float64, Float64, Float64}, NTuple{4, Float64}}

julia> @allocated ext._beta_inc_grad(1.2, 2.3, 0.4)
32
```

We have allocation tests in the suite already, so an `@inferred`/`@allocated` check on `_beta_inc_grad` would be worth adding.

### `beta_inc_inv` rule doesn't promote its arguments

The two `beta_inc` rules use `map(float, promote(a, b, x))` but the `beta_inc_inv` one calls `_beta_inc_grad(a, b, x)` directly, so anything but `Float64` arguments fails:

```julia
julia> frule((NoTangent(), 1.0, 0.0, 0.0), beta_inc_inv, 1, 2, 0.5)
ERROR: MethodError: no method matching _beta_inc_grad(::Int64, ::Int64, ::Float64)
```

`beta_inc_inv(1, 2, 0.5)` itself works fine, so this is a regression introduced by the rule.

### `Float16` isn't supported although the primal supports it

`_beta_inc_grad` is restricted to `Union{Float64, Float32}`, but `_beta_inc` has a `Union{Float16, Float32}` method (`src/beta_inc.jl:908`) that computes in `Float64` and converts back. Could you follow that pattern here? It fixes `Float16`, avoids compiling the whole CF a second time for `Float32`, and should let you tighten the `rtol=5e-4` in the `Float32` tests considerably.

### Unconverged results are returned with just a `@warn`

`maxapp = 200` isn't enough for large parameters, and the result can be quite wrong:

| `a`, `b`, `x` | rel. error of `∂I/∂a` | |
|---|---|---|
| `1000, 1000, 0.5` | 1.5e-12 | converged |
| `1e4, 1e4, 0.5` | 2.4e-10 | converged |
| `1e5, 1e5, 0.5` | 2.5e-8 | warns |
| `1e6, 1e6, 0.5` | **2.3e-2** | warns |

A 2% gradient isn't something a caller can be expected to notice. I'd either increase `maxapp`, or return `NaN`, or at the very least document the limitation. And the `@warn` needs `maxlog=1` — as is it will be emitted on every gradient evaluation inside an optimiser or sampler. Note that `@test_logs` in `test/chainrules.jl:290` pins the exact message, so that test would have to be adjusted along with it.

### `NaN`s for degenerate arguments

```julia
julia> beta_inc(0.0, 2.0, 0.4), ext._beta_inc_grad(0.0, 2.0, 0.4)
((1.0, 0.0), (NaN, NaN, 0.0))

julia> frule((NoTangent(), 0.0, 0.0, 1.0), beta_inc_inv, 1.2, 2.3, 1.0)[2]
Tangent{Tuple{Float64, Float64}}(NaN, NaN)
```

Same for `b == 0` and for `beta_inc_inv` at `p == 0`. The `beta_inc_inv` ones come from `0 * inv(0)` in `-dIa * inv_dIx`, partly aggravated by the `x == 1` issue above. The primal handles all of these, so I think the rules should too, even if the answer is a deliberate `NaN`.

### `sign(Bn)` in the guard against small `Bn`

```julia
invBn = (Bn > tiny || Bn < -tiny) && isfinite(Bn) ? inv(Bn) : inv(sign(Bn) * tiny)
```

`sign(0.0) == 0.0`, so for `Bn == 0` — the case the guard is there for — this returns `Inf`. `inv(copysign(tiny, Bn))` would do what you want.

### Smaller things

- `_dan_dq` takes `q` and `da1_dq` and uses neither, and its comment claims it returns the precomputed `∂a₁/∂q` for `n == 1`, which it doesn't. Same for the `n == 1` remark in `_dan_dp`. The loop starts at `n = 2`, so I think all the `n == 1` comments in these helpers can go, together with the unused arguments.
- `_dK_dq` is missing the `return`, unlike `_dK_dp` right above it.
- `An, an, Bn = _nextapp1(f, p, q)` assigns the returned `bn` to `Bn`. It's correct since `B₁ == b₁`, but it reads like a bug — maybe just return `(An, Bn)`?
- `ϵ = eps(T)*T(1e4)` with the `# 0) Previously keyword arguments:` comment looks like leftover. Either make it a keyword like `maxapp`/`minapp` or drop the comment. Also the step numbering skips 2.
- `_dbn_dp` and `_dbn_dq` recompute `A`, `N1`, `N` and `D` identically and are both called every iteration.
- `pfq_2` (`pf/q + 2`) and `pf_2q` (`pf + 2q`) differing by one character is going to bite someone eventually.
- Mixed `2*n` / `2p` / `2q` spacing, and a double blank line between `_dan_dp` and `_dan_dq`.
- The PR description cites Boik & Robinson-Cox 1998 (CSDA 27(1)) while the comment cites 1999 (JSS 3(1)). Both exist, but probably pick one. Keeping the reference in the comments is fine IMO, that's what we do in `src/beta_inc.jl` as well.
- The code says it's adapted from `arzwa/IncBetaDer.jl` — could you check what licence that is under and whether a comment is enough?

### Tests

The grid over `(a, b, x)` is good and it only adds ~30s, so no complaints there. But:

- Nothing near `x -> 0` or `x -> 1`, which is exactly where the bug above is.
- No `Integer` arguments, no `Float16`.
- `test_frule`/`test_rrule` are called without a surrounding `@testset`, so a failure gives you no idea which `(a, b, x)` it was. Could you wrap the loop bodies in `@testset "a=$a b=$b x=$x"`?

### Other

`beta_inc_inv(a, b, p, q)` doesn't get a rule, although its 4-argument form is documented next to `beta_inc(a, b, x, y)` in the overview table. Could you add it with the same symmetric split?

Would you mind squashing/rewording the commits before merge?

lrnv committed:
Exact chainrules derivatives for `beta_inc` and `beta_inc_inv`

## Summary

This PR adds ChainRules coverage for the regularized incomplete beta function and its inverse.

- Adds analytic `frule`/`rrule` definitions for:
  - `beta_inc(a, b, x) -> (p, q)`
  - `beta_inc(a, b, x, y)`, with `x + y = 1`
  - `beta_inc_inv(a, b, p) -> (x, 1 - x)`
  - `beta_inc_inv(a, b, p, q)`, with `p + q = 1`
- Uses symmetric tangent/cotangent splits for the redundant four-argument parametrizations.
- Supports integer inputs through promotion and supports `Float16`/`Float32` by computing the differentiated continued fraction in `Float64` and converting back.
- Handles endpoints and degenerate shape parameters explicitly.
- Adds extensive tests for forward and reverse rules, endpoint behavior, type stability, allocations, promotion, non-convergence, and derivatives along the four-argument constraints.

## Implementation

The parameter derivatives of `beta_inc` are computed by differentiating the continued-fraction method described by Boik & Robinson-Cox (1999). This is **not a translation of the S+ source code**: the implementation is independently adapted from [arzwa/IncBetaDer](https://github.com/arzwa/IncBetaDer), now MIT-licensed, whose author also [explicitly permitted reuse](https://github.com/JuliaMath/SpecialFunctions.jl/pull/506#issuecomment-3369042219).

The continued fraction exits as soon as its values converge. Its upper bound is `maxapp = 10_000`, rather than the historical value of 200, because difficult nearly symmetric cases with large shape parameters require substantially more approximants. If convergence is not reached, the helper emits a rate-limited warning and returns `NaN` derivatives instead of silently returning an inaccurate gradient.

Small recurrence helpers used only once have been integrated into the loop. The remaining shared helper stays `@inline` because it is called four times per iteration and benchmarking showed a material performance benefit.

## Motivation

This allows automatic differentiation through `beta_inc` and `beta_inc_inv`. Missing derivatives affect, among other use cases:

- beta quantiles and beta-distribution marginals;
- Student-t and F distribution workflows built on incomplete beta functions;
- fitting `Distributions.MvTDist`;
- applications in packages such as Copulas.jl;
- the use case discussed in [this Julia Discourse thread](https://discourse.julialang.org/t/gaussian-copula-priors-in-turing-auto-differentiation-error-with-beta-quantile-function/132779/5).

These rules naturally belong in SpecialFunctions.jl's ChainRulesCore extension.

## Validation

The tests cover a broad grid of `(a, b, x)`, including values close to `x = 0` and `x = 1`, integer and reduced-precision inputs, and the constrained four-argument APIs. Sensitive large-parameter regression cases include `a = b = 10^6` and `a = b = 10^8` at `x = 0.5`; these exercise cases that need more than 200 and, for `10^8`, thousands of approximants.

The full SpecialFunctions.jl test suite passes, and the new code is covered by the test suite.

## Reference

> Boik, R. J., & Robinson-Cox, J. F. (1999). Derivatives of the incomplete beta function. *Journal of Statistical Software, 3*(1). https://doi.org/10.18637/jss.v003.i01
